File: 6_last_digit_of_the_sum_of_fibonacci_numbers/FibonacciSumLastDigit.java
/*
 * Author: Jehadur Rahman
 * Organization: CyArm
 * GitHub: https://github.com/JehadurRE
 * Website: jehadurre.me
 * Date: 2025-07-10 09:43 (Thu)
 */

import java.util.*;

public class FibonacciSumLastDigit {
    private static long getFibonacciSumNaive(long n) {
        if (n <= 1)
            return n;

        long previous = 0;
        long current = 1;
        long sum = 1;

        for (long i = 0; i < n - 1; ++i) {
            long tmp_previous = previous;
            previous = current;
            current = tmp_previous + current;
            sum += current;
        }

        return sum % 10;
    }

    private static int fibonacciSumFast(long n) {
        n = n % 60; // Pisano period for mod 10

        if (n <= 1)
            return (int) n;

        long previous = 0;
        long current = 1;
        long sum = 1;

        for (long i = 0; i < n - 1; ++i) {
            long tmp_previous = previous;
            previous = current;
            current = (tmp_previous + current) % 10;
            sum = (sum + current) % 10;
        }

        return (int) sum;
    }
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        long n = scanner.nextLong();
        // long s = getFibonacciSumNaive(n);
        int s = fibonacciSumFast(n);
        System.out.println(s);
    }
}

File: 6_last_digit_of_the_sum_of_fibonacci_numbers/FibonacciSumLastDigit.kt
/*
 * Author: Jehadur Rahman
 * Organization: CyArm
 * GitHub: https://github.com/JehadurRE
 * Website: jehadurre.me
 * Date: 2025-07-10 09:43 (Thu)
 */

import java.util.*

fun getFibonacciSumNaive(n: Long): Long {
    if (n <= 1)
        return n

    var previous: Long = 0
    var current: Long = 1
    var sum: Long = 1

    for (i in 0 until n - 1) {
        val tmpPrevious = previous
        previous = current
        current += tmpPrevious
        sum += current
    }

    return sum % 10
}

fun fibonacciSumFast(n: Long): Int {
    val newN = n % 60 // Pisano period for mod 10

    if (newN <= 1)
        return newN.toInt()

    var previous: Long = 0
    var current: Long = 1
    var sum: Long = 1

    for (i in 0 until newN - 1) {
        val tmpPrevious = previous
        previous = current
        current = (tmpPrevious + current) % 10
        sum = (sum + current) % 10
    }

    return sum.toInt()
}

fun main(args: Array<String>) {
    val scanner = Scanner(System.`in`)
    val n = scanner.nextLong()
    // val s = getFibonacciSumNaive(n)
    val s = fibonacciSumFast(n)
    println(s)
}